subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c

```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+{
+	size_t block_cnt;
+	uint8_t aes_input[FP_CRYPTO_AES128_BLOCK_LEN];
+	uint8_t aes_block[FP_CRYPTO_AES128_BLOCK_LEN];
+	int err;
+
+	if (data_len % FP_CRYPTO_AES128_BLOCK_LEN == 0) {
+		block_cnt = data_len / FP_CRYPTO_AES128_BLOCK_LEN;
+	} else {
+		block_cnt = (data_len / FP_CRYPTO_AES128_BLOCK_LEN) + 1;
+	}
+
+	memset(&aes_input[AES_INPUT_PAD_START], 0, AES_INPUT_PAD_LEN);
+	memcpy(&aes_input[AES_INPUT_NONCE_START], nonce, FP_CRYPTO_NONCE_LEN);
+
+	memcpy(out, in, data_len);
+	for (size_t i = 0; i < block_cnt; i++) {
+		aes_input[0] = (uint8_t)i;
+		err = fp_crypto_aes128_encrypt(aes_block, aes_input, k);
+		if (err) {
+			return err;
+		}
+
+		for (size_t j = 0; (j < FP_CRYPTO_AES128_BLOCK_LEN) &&
+				   (i * FP_CRYPTO_AES128_BLOCK_LEN + j < data_len); j++) {
+			out[i * FP_CRYPTO_AES128_BLOCK_LEN + j] ^= aes_block[j];
+		}
+	}
+
+	return 0;
+}
+
+int fp_crypto_aes_ctr_decrypt(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *k,
+			      const uint8_t *nonce)
+{
+	return fp_crypto_aes_ctr_encrypt(out, in, data_len, k, nonce);
+}
+
+int fp_crypto_hmac_sha256(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *k)
+{
+	uint8_t inner_sha_input[HMAC_SHA_PAD_LEN + data_len];
+	uint8_t sha_input[HMAC_SHA_PAD_LEN + FP_CRYPTO_SHA256_HASH_LEN];
+	int err;
+
+	for (size_t i = 0; i < FP_CRYPTO_AES128_BLOCK_LEN; i++) {
+		inner_sha_input[i] = k[i] ^ HMAC_SHA_IPAD_VAL;
+	}
+	memset(&inner_sha_input[FP_CRYPTO_AES128_BLOCK_LEN], HMAC_SHA_IPAD_VAL,
+	       HMAC_SHA_PAD_LEN - FP_CRYPTO_AES128_BLOCK_LEN);
+	memcpy(&inner_sha_input[INNER_SHA_DATA_START], in, data_len);
+	err = fp_crypto_sha256(&sha_input[HMAC_SHA_INNER_SHA_START], inner_sha_input,
+			       sizeof(inner_sha_input));
+	if (err) {
+		return err;
+	}
+
+	for (size_t i = 0; i < FP_CRYPTO_AES128_BLOCK_LEN; i++) {
+		sha_input[i] = k[i] ^ HMAC_SHA_OPAD_VAL;
+	}
+	memset(&sha_input[FP_CRYPTO_AES128_BLOCK_LEN], HMAC_SHA_OPAD_VAL,
+	       HMAC_SHA_PAD_LEN - FP_CRYPTO_AES128_BLOCK_LEN);
+	return fp_crypto_sha256(out, sha_input, sizeof(sha_input));
+}
+
+size_t fp_crypto_additional_data_packet_size(size_t data_len)
+{
+	return ADDITIONAL_DATA_SHA_LEN + FP_CRYPTO_NONCE_LEN + data_len;
+}
+
+int fp_crypto_additional_data_encode(uint8_t *out, const uint8_t *data, size_t data_len,
+				     const uint8_t *k, const uint8_t *nonce)
+{
+	int err;
+	uint8_t hmac_sha256_buf[FP_CRYPTO_SHA256_HASH_LEN];
+
+	memcpy(&out[ADDITIONAL_DATA_NONCE_START], nonce, FP_CRYPTO_NONCE_LEN);
+	err = fp_crypto_aes_ctr_encrypt(&out[ADDITIONAL_DATA_ENC_DATA_START], data, data_len, k,
+					nonce);
+	if (err) {
+		return err;
+	}
+
+	err = fp_crypto_hmac_sha256(hmac_sha256_buf, &out[ADDITIONAL_DATA_NONCE_START],
+				    FP_CRYPTO_NONCE_LEN + data_len, k);
+	if (err) {
+		return err;
+	}
+
+	memcpy(out, hmac_sha256_buf, ADDITIONAL_DATA_SHA_LEN);
+
+	return 0;
+}
+
+size_t fp_crypto_additional_data_decoded_size(size_t packet_len)
+{
+	__ASSERT_NO_MSG(packet_len >= ADDITIONAL_DATA_ENC_DATA_START);
```

Comment:
It seems to be API function, should it fails with assertion or return an error, for example 0 when packet_len is to small. In case where assertion will be disabled this function return unexpected value if packet_len will be to small. Consider adding documentation about minimum length of the packet and you can return 0 in case when packet_len is to small as a protection in case when assertion are disabled